ZFS panics in abd_cmp

<!-- Please fill out the following template, which will help other contributors address your issue. -->

<!--
Thank you for reporting an issue.

*IMPORTANT* - Please check our issue tracker before opening a new issue.
Additional valuable information can be found in the OpenZFS documentation
and mailing list archives.

Please fill in as much of the template as possible.
-->

### System information
<!--  add version after "|" character -->
Type | Version/Name
 --- | ---
Distribution Name	| FreeBSD
Distribution Version	| 15.1-STABLE
Kernel Version	| 1501501
Architecture	| amd64
OpenZFS Version	| 2.4.2-1-FreeBSD_g6330a45b0
<!--
Command to find OpenZFS version:
zfs version

Commands to find kernel version:
uname -r		# Linux
freebsd-version -r	# FreeBSD
-->

### Describe the problem you're observing
System panics. We observed the same panic on different systems with the same version.

### Describe how to reproduce the problem
It seems it is related to deduplication.
```
# zfs get dedup zroot
NAME   PROPERTY  VALUE          SOURCE
zroot  dedup     verify         received
```

The problem was observed after upgrade from:
```
% zfs version
zfs-2.3.0-rc1-FreeBSD_g3a9fca901
zfs-kmod-2.3.0-rc1-FreeBSD_g3a9fca901
```

### Include any warning/errors/backtraces from the system logs
<!--
*IMPORTANT* - Please mark logs and text output from terminal commands
or else Github will not display them correctly.
An example is provided below.

Example:
```
this is an example how log text should be marked (wrap it with ```)
```
-->

```
Fatal trap 12: page fault while in kernel mode
cpuid = 1; apic id = 02
fault virtual address   = 0x8
fault code              = supervisor read data, page not present
instruction pointer     = 0x20:0xffffffff8253cca7
stack pointer           = 0x28:0xfffffe00fb413bb0
frame pointer           = 0x28:0xfffffe00fb413bb0
code segment            = base 0x0, limit 0xfffff, type 0x1b
                        = DPL 0, pres 1, long 1, def32 0, gran 1
processor eflags        = interrupt enabled, resume, IOPL = 0
current process         = 0 (z_wr_iss_0)
rdi: fffff8036f23a138 rsi: 0000000000000000 rdx: 0000000000000000
rcx: 0000000000020000  r8: 0000000000000000  r9: 0000000000000000
rax: 0000000000000008 rbx: fffffe00fb413bf8 rbp: fffffe00fb413bb0
r10: 000000002d1c6717 r11: 00000000e0d087f7 r12: 0000000000000000
r13: 0000000000020000 r14: 0000000000020000 r15: fffffe00fb413bc0
trap number             = 12
panic: page fault
cpuid = 1
time = 1782453811
KDB: stack backtrace:
#0 0xffffffff80bd609d at kdb_backtrace+0x5d
#1 0xffffffff80b86be9 at vpanic+0x149
#2 0xffffffff80b86a93 at panic+0x43
#3 0xffffffff8109ac94 at trap_pfault+0x384
#4 0xffffffff81071058 at calltrap+0x8
#5 0xffffffff82579b25 at abd_iterate_func2+0x2e5
#6 0xffffffff82579bc4 at abd_cmp+0x24
#7 0xffffffff8271df39 at zio_ddt_write+0x5d9
#8 0xffffffff8271a208 at zio_execute+0x38
#9 0xffffffff80becd7e at taskqueue_run_locked+0x18e
#10 0xffffffff80bedf42 at taskqueue_thread_loop+0xc2
#11 0xffffffff80b3cf7b at fork_exit+0x7b
#12 0xffffffff8107207e at fork_trampoline+0xe
Uptime: 7d10h3m10s

(kgdb) bt
#0  __curthread () at /usr/src/sys/amd64/include/pcpu_aux.h:57
#1  doadump (textdump=<optimized out>) at /usr/src/sys/kern/kern_shutdown.c:399
#2  0xffffffff80b8674b in kern_reboot (howto=260) at /usr/src/sys/kern/kern_shutdown.c:519
#3  0xffffffff80b86c4d in vpanic (fmt=<optimized out>, ap=ap@entry=0xfffffe00fb413a70) at /usr/src/sys/kern/kern_shutdown.c:974
#4  0xffffffff80b86a93 in panic (fmt=<unavailable>) at /usr/src/sys/kern/kern_shutdown.c:887
#5  0xffffffff8109ac94 in trap_fatal (frame=<optimized out>, eva=<optimized out>) at /usr/src/sys/amd64/amd64/trap.c:1028
#6  0xffffffff8109ac94 in trap_pfault (frame=0xfffffe00fb413af0, usermode=false, signo=<optimized out>, ucode=<optimized out>)
#7  <signal handler called>
#8  list_next (list=list@entry=0xfffff8036f23a138, object=object@entry=0x0) at /usr/src/sys/contrib/openzfs/module/os/freebsd/spl/list.c:173
#9  0xffffffff82579b25 in abd_advance_abd_iter (abd=0xfffff8036f23a100, cabd=0x0, aiter=0xfffffe00fb413bf8, len=131072) at /usr/src/sys/contrib/openzfs/module/zfs/abd.c:752
#10 abd_iterate_func2 (dabd=0xfffff8000f1a8e80, sabd=0xfffff8036f23a100, doff=doff@entry=0, soff=soff@entry=0, size=<optimized out>, func=0xffffffff82579bd0 <abd_cmp_cb>, private=0x0)
    at /usr/src/sys/contrib/openzfs/module/zfs/abd.c:980
#11 0xffffffff82579bc4 in abd_cmp (dabd=0xfffff8036f23a138, sabd=0x0) at /usr/src/sys/contrib/openzfs/module/zfs/abd.c:1018
#12 0xffffffff8271df39 in zio_ddt_collision (zio=0xfffffe01301c8000, ddt=0xfffffe0102aeb000, dde=0xfffff8005c1a9958) at /usr/src/sys/contrib/openzfs/module/zfs/zio.c:3606
#13 zio_ddt_write (zio=0xfffffe01301c8000) at /usr/src/sys/contrib/openzfs/module/zfs/zio.c:3815
#14 0xffffffff8271a208 in __zio_execute (zio=<optimized out>) at /usr/src/sys/contrib/openzfs/module/zfs/zio.c:2485
#15 zio_execute (zio=<optimized out>) at /usr/src/sys/contrib/openzfs/module/zfs/zio.c:2396
#16 0xffffffff80becd7e in taskqueue_run_locked (queue=queue@entry=0xfffff800030a8900) at /usr/src/sys/kern/subr_taskqueue.c:517
#17 0xffffffff80bedf42 in taskqueue_thread_loop (arg=arg@entry=0xfffff80001a7bce0) at /usr/src/sys/kern/subr_taskqueue.c:829
#18 0xffffffff80b3cf7b in fork_exit (callout=0xffffffff80bede80 <taskqueue_thread_loop>, arg=0xfffff80001a7bce0, frame=0xfffffe00fb413f40) at /usr/src/sys/kern/kern_fork.c:1202
#19 <signal handler called>

(kgdb) bt full
M#0  __curthread () at /usr/src/sys/amd64/include/pcpu_aux.h:57
        td = <optimized out>
#1  doadump (textdump=<optimized out>) at /usr/src/sys/kern/kern_shutdown.c:399
        error = 0
        coredump = <optimized out>
#2  0xffffffff80b8674b in kern_reboot (howto=260) at /usr/src/sys/kern/kern_shutdown.c:519
        once = 0
#3  0xffffffff80b86c4d in vpanic (fmt=<optimized out>, ap=ap@entry=0xfffffe00fb413a70) at /usr/src/sys/kern/kern_shutdown.c:974
        buf = "page fault", '\000' <repeats 245 times>
        __pc = 0x0
        __pc = 0x0
        __pc = 0x0
        other_cpus = {__bits = {13, 0 <repeats 15 times>}}
        td = 0xfffff8010b0dc780
        bootopt = <unavailable>
        newpanic = <optimized out>
#4  0xffffffff80b86a93 in panic (fmt=<unavailable>) at /usr/src/sys/kern/kern_shutdown.c:887
        ap = {{gp_offset = 16, fp_offset = 48, overflow_arg_area = 0xfffffe00fb413aa0, reg_save_area = 0xfffffe00fb413a40}}
#5  0xffffffff8109ac94 in trap_fatal (frame=<optimized out>, eva=<optimized out>) at /usr/src/sys/amd64/amd64/trap.c:1028
        type = <optimized out>
        handled = <optimized out>
#6  0xffffffff8109ac94 in trap_pfault (frame=0xfffffe00fb413af0, usermode=false, signo=<optimized out>, ucode=<optimized out>)
        __pc = 0x0
        __pc = 0x0
        __pc = 0x0
        td = <optimized out>
        p = <optimized out>
        eva = <optimized out>
        map = <optimized out>
        ftype = <optimized out>
        rv = <optimized out>
#7  <signal handler called>
No locals.
#8  list_next (list=list@entry=0xfffff8036f23a138, object=object@entry=0x0) at /usr/src/sys/contrib/openzfs/module/os/freebsd/spl/list.c:173
        node = 0x8
#9  0xffffffff82579b25 in abd_advance_abd_iter (abd=0xfffff8036f23a100, cabd=0x0, aiter=0xfffffe00fb413bf8, len=131072) at /usr/src/sys/contrib/openzfs/module/zfs/abd.c:752
No locals.
#10 abd_iterate_func2 (dabd=0xfffff8000f1a8e80, sabd=0xfffff8036f23a100, doff=doff@entry=0, soff=soff@entry=0, size=<optimized out>, func=0xffffffff82579bd0 <abd_cmp_cb>, private=0x0)
    at /usr/src/sys/contrib/openzfs/module/zfs/abd.c:980
        len = 131072
        slen = <optimized out>
        dlen = <optimized out>
        daiter = {{{iter_mapaddr = 0x0, iter_mapsize = 0}, {iter_page = 0x0, iter_page_doff = 0, iter_page_dsize = 0}}, iter_abd = 0xfffff8000f1a8e80, iter_pos = 131072, iter_offset = 0,
          sf = 0x0}
        saiter = {{{iter_mapaddr = 0x0, iter_mapsize = 0}, {iter_page = 0x0, iter_page_doff = 0, iter_page_dsize = 0}}, iter_abd = 0xfffff8036f23a100, iter_pos = 131072, iter_offset = 0,
          sf = 0x0}
        ret = <optimized out>
        c_dabd = 0x0
        c_sabd = 0x0
#11 0xffffffff82579bc4 in abd_cmp (dabd=0xfffff8036f23a138, sabd=0x0) at /usr/src/sys/contrib/openzfs/module/zfs/abd.c:1018
No locals.
#12 0xffffffff8271df39 in zio_ddt_collision (zio=0xfffffe01301c8000, ddt=0xfffffe0102aeb000, dde=0xfffff8005c1a9958) at /usr/src/sys/contrib/openzfs/module/zfs/zio.c:3606
        lio = 0x8
        p = <optimized out>
        spa = <optimized out>
        do_raw = <optimized out>
#13 zio_ddt_write (zio=0xfffffe01301c8000) at /usr/src/sys/contrib/openzfs/module/zfs/zio.c:3815
        czp = {zp_checksum = 128, zp_compress = 199, zp_complevel = 13 '\r', zp_level = 11 '\v', zp_copies = 1 '\001', zp_gang_copies = 248 '\370', zp_type = 255, zp_storage_type = 255,
          zp_dedup = 0, zp_dedup_verify = 0, zp_nopwrite = 0, zp_brtwrite = 0, zp_encrypt = 0, zp_byteorder = 0, zp_direct_write = 0, zp_rewrite = 0, zp_zpl_smallblk = 5,
          zp_salt = "\200\216t\327\a\370\377\377", zp_iv = "\000\000\000\000\000\000\000\000\000\240]\002", zp_mac = "\001\376\377\377X\211\n\003\000\370\377\377\000\000\000"}
        txg = 129458
        spa = 0xfffffe01025da000
        bp = 0xfffff80418d00a98
        zp = 0xfffffe01301c8020
        ddt = 0xfffffe0102aeb000
        dde = 0xfffff8005c1a9958
        p = <optimized out>
        v = <optimized out>
        ddp = <optimized out>
        have_dvas = <optimized out>
        is_ganged = <optimized out>
        parent_dvas = <optimized out>
        dde_io = <optimized out>
        cio = <optimized out>
        need_dvas = <optimized out>
#14 0xffffffff8271a208 in __zio_execute (zio=<optimized out>) at /usr/src/sys/contrib/openzfs/module/zfs/zio.c:2485
        stage = ZIO_STAGE_DDT_WRITE
        pipeline = <optimized out>
        cut = <optimized out>
#15 zio_execute (zio=<optimized out>) at /usr/src/sys/contrib/openzfs/module/zfs/zio.c:2396
        cookie = 0
#16 0xffffffff80becd7e in taskqueue_run_locked (queue=queue@entry=0xfffff800030a8900) at /usr/src/sys/kern/subr_taskqueue.c:517
        et = {et_link = {tqe_next = 0x1, tqe_prev = 0x0}, et_td = 0xffffffff8127339b, et_section = {bucket = 0}, et_old_priority = 0 '\000'}
        tb = {tb_running = 0xfffffe01301c83b0, tb_seq = 22617674, tb_canceling = false, tb_link = {le_next = 0x0, le_prev = 0xfffffe00fb40ee88}}
        in_net_epoch = false
        pending = 1
        task = <optimized out>
#17 0xffffffff80bedf42 in taskqueue_thread_loop (arg=arg@entry=0xfffff80001a7bce0) at /usr/src/sys/kern/subr_taskqueue.c:829
        tqp = <optimized out>
        tq = 0xfffff800030a8900
#18 0xffffffff80b3cf7b in fork_exit (callout=0xffffffff80bede80 <taskqueue_thread_loop>, arg=0xfffff80001a7bce0, frame=0xfffffe00fb413f40) at /usr/src/sys/kern/kern_fork.c:1202
        __pc = 0x0
        __pc = 0x0
        td = 0xfffff8010b0dc780
        p = 0xffffffff81b82c70 <proc0>
        dtd = <optimized out>
#19 <signal handler called>
```

--- a/module/zfs/zio.c
+++ b/module/zfs/zio.c
@@ -3587,16 +3587,26 @@ zio_ddt_collision(zio_t *zio, ddt_t *ddt, ddt_entry_t *dde)
 		if (dde->dde_io == NULL)
 			continue;
 
+		/*
+		 * Lock dde_io to prevent the lead zio from completing
+		 * and freeing its ABD while we compare against it.
+		 */
+		mutex_enter(&dde->dde_io->dde_io_lock);
 		zio_t *lio = dde->dde_io->dde_lead_zio[p];
-		if (lio == NULL)
+		if (lio == NULL) {
+			mutex_exit(&dde->dde_io->dde_io_lock);
 			continue;
-
-		if (do_raw)
-			return (lio->io_size != zio->io_size ||
-			    abd_cmp(zio->io_abd, lio->io_abd) != 0);
-
-		return (lio->io_orig_size != zio->io_orig_size ||
-		    abd_cmp(zio->io_orig_abd, lio->io_orig_abd) != 0);
+		}
+		boolean_t collision;
+		if (do_raw) {
+			collision = lio->io_size != zio->io_size ||
+			    abd_cmp(zio->io_abd, lio->io_abd) != 0;
+		} else {
+			collision = lio->io_orig_size != zio->io_orig_size ||
+			    abd_cmp(zio->io_orig_abd, lio->io_orig_abd) != 0;
+		}
+		mutex_exit(&dde->dde_io->dde_io_lock);
+		return (collision);
 	}
 
 	for (int p = 0; p < DDT_NPHYS(ddt); p++) {

